runtime: crash in split by func

Missing validation of value returned by function used in splitByFn.

```sh
λ> a = "hello world"
λ> fn b() { echo "oops" }
λ> c <= split($a, $b)
oops
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x53054d]

goroutine 55 [running]:
panic(0x610200, 0xc42000c160)
	/home/i4k/.gvm/gos/go1.7/src/runtime/panic.go:500 +0x1a1
github.com/NeowayLabs/nash/internal/sh.splitByFn.func1(0xc400000068, 0xb)
	/home/i4k/.gvm/pkgsets/go1.7/global/src/github.com/NeowayLabs/nash/internal/sh/split.go:162 +0x17d
strings.FieldsFunc(0xc4203e0330, 0xb, 0xc42003be90, 0xc4201043a0, 0x0, 0x0)
	/home/i4k/.gvm/gos/go1.7/src/strings/strings.go:315 +0x7a
github.com/NeowayLabs/nash/internal/sh.splitByFn(0xc4203e0330, 0xb, 0x934f60, 0xc4201b79f0, 0x0, 0x0, 0x0)
	/home/i4k/.gvm/pkgsets/go1.7/global/src/github.com/NeowayLabs/nash/internal/sh/split.go:173 +0x5f
github.com/NeowayLabs/nash/internal/sh.(*SplitFn).run(0xc420084100, 0x0, 0x0)
	/home/i4k/.gvm/pkgsets/go1.7/global/src/github.com/NeowayLabs/nash/internal/sh/split.go:58 +0x39a
github.com/NeowayLabs/nash/internal/sh.(*SplitFn).Start.func1(0xc420084100)
	/home/i4k/.gvm/pkgsets/go1.7/global/src/github.com/NeowayLabs/nash/internal/sh/split.go:77 +0x2f
created by github.com/NeowayLabs/nash/internal/sh.(*SplitFn).Start
	/home/i4k/.gvm/pkgsets/go1.7/global/src/github.com/NeowayLabs/nash/internal/sh/split.go:79 +0x75
ERROR: exit status 2
```

--- a/internal/sh/builtin/split_test.go
+++ b/internal/sh/builtin/split_test.go
@@ -45,14 +45,9 @@ func TestSplit(t *testing.T) {
 	for name, desc := range tests {
 		t.Run(name, func(t *testing.T) {
 			var output bytes.Buffer
-			shell, err := nash.New()
-
-			if err != nil {
-				t.Fatalf("unexpected err: %s", err)
-			}
-
+			shell := newShell(t)
 			shell.SetStdout(&output)
-			err = shell.ExecFile(desc.script, "mock cmd name", desc.word, desc.sep)
+			err := shell.ExecFile(desc.script, "mock cmd name", desc.word, desc.sep)
 
 			if err != nil {
 				t.Fatalf("unexpected err: %s", err)
@@ -63,4 +58,47 @@ func TestSplit(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestSplitingByFuncWrongWontPanic(t *testing.T) {
+	// Regression for: https://github.com/NeowayLabs/nash/issues/177
+
+	badscripts := map[string]string{
+		"noReturnValue": `
+			fn b() { echo "oops" }
+			split("lalala", $b)
+		`,
+		"returnValueIsList": `
+			fn b() { return () }
+			split("lalala", $b)
+		`,
+		"returnValueIsFunc": `
+			fn b() { 
+				fn x () {}
+				return $x
+			}
+			split("lalala", $b)
+		`,
+	}
+
+	for testname, badscript := range badscripts {
+
+		t.Run(testname, func(t *testing.T) {
+			shell := newShell(t)
+			_, err := shell.ExecOutput("whatever", badscript)
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+func newShell(t *testing.T) *nash.Shell {
+	shell, err := nash.New()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return shell
 }

